runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/ArtifactStagingService.java

```diff
@@ -0,0 +1,602 @@
+          throw new RuntimeException(exn);
+        }
+        // Limit to the last contiguous alpha-numeric sequence. In particular, this will exclude
+        // all path separators.
+        List<String> components = Splitter.onPattern("[^A-Za-z-_.]]").splitToList(path);
+        String base = components.get(components.size() - 1);
+        return clip(
+            String.format("%s-%s-%s", idGenerator.getId(), clip(environment, 25), base), 100);
+      }
+
+      private String clip(String s, int maxLength) {
+        return s.length() < maxLength ? s : s.substring(0, maxLength);
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+        stagingExecutor.shutdownNow();
+        LOG.error("Error staging artifacts", throwable);
+        state = State.ERROR;
+      }
+
+      @Override
+      public void onCompleted() {
+        Preconditions.checkArgument(state == State.DONE);
+      }
+    };
+  }
+
+  @Override
+  public void close() throws Exception {
+    // Nothing to close.
+  }
+
+  /**
+   * Lazily stages artifacts by letting an ArtifactStagingService resolve and request artifacts.
+   *
+   * @param retrievalService an ArtifactRetrievalService used to resolve and retrieve artifacts
+   * @param stagingService an ArtifactStagingService stub which will request artifacts
+   * @param stagingToken the staging token of the job whose artifacts will be retrieved
+   * @throws InterruptedException
+   * @throws IOException
+   */
+  public static void offer(
+      ArtifactRetrievalService retrievalService,
+      ArtifactStagingServiceGrpc.ArtifactStagingServiceStub stagingService,
+      String stagingToken)
+      throws ExecutionException, InterruptedException {
+    new StagingDriver(retrievalService, stagingService, stagingToken).getCompletionFuture().get();
+  }
+
+  /** Actually implements the reverse retrieval protocol. */
+  private static class StagingDriver implements StreamObserver<ArtifactApi.ArtifactRequestWrapper> {
+
+    private final ArtifactRetrievalService retrievalService;
+    private final StreamObserver<ArtifactApi.ArtifactResponseWrapper> responseObserver;
+    private final CompletableFuture<Void> completionFuture;
+
+    public StagingDriver(
+        ArtifactRetrievalService retrievalService,
+        ArtifactStagingServiceGrpc.ArtifactStagingServiceStub stagingService,
+        String stagingToken) {
+      this.retrievalService = retrievalService;
+      responseObserver = stagingService.reverseArtifactRetrievalService(this);
+      responseObserver.onNext(
+          ArtifactApi.ArtifactResponseWrapper.newBuilder().setStagingToken(stagingToken).build());
+      completionFuture = new CompletableFuture<Void>();
+    }
+
+    public CompletableFuture<?> getCompletionFuture() {
+      return completionFuture;
+    }
+
+    @Override
+    public void onNext(ArtifactApi.ArtifactRequestWrapper requestWrapper) {
+      if (requestWrapper.hasResolveArtifact()) {
```

Comment:
if the future is completed, we should call responseObserver.onError with the exception

This allows us to propagate the root cause further along.